proposals/0030-core-image-story.md

```diff
@@ -0,0 +1,427 @@
+
+# Nomenclature
+
+- **CLS**: Content Layout Shift. A score determining how much content shifted on your page during loading, you want to be as close to 0 (or "no CLS") as possible.
+
+# Example
+
+> Note: All the `src` paths in the examples below represent the final paths at build. During development, those paths may be different due to implementations details.
+
+> Note 2: All the results below, unless specified otherwise, are done using the default image services provided with Astro.
+
+## Image Component
+
+### Image located in `src`, in an Astro file
+
+```astro
+---
+import { Image } from 'astro:assets';
+import myImage from "../assets/my_image.png"; // Image is 1600x900
+---
+
+<Image src={myImage} alt="..." />
+```
+
+#### Result
+
+```html
+<!-- Image is optimized, proper attributes are enforced -->
+<img
+  src="/_astro/my_image.hash.webp"
+  width="1600"
+  height="900"
+  decoding="async"
+  loading="lazy"
+  alt="..."
+/>
+```
+
+### Remote images
+
+```astro
+---
+import { Image } from "astro:assets";
+// Image is 1920x1080
+---
+
+<Image src="https://example.com/image.png" alt="..." />
+<!-- ERROR! `width` and `height` are required -->
+
+<Image src="https://example.com/image.png" width={1280} alt="..." />
+<!-- ERROR! `height` is required -->
+
+<Image src="https://example.com/image.png" width={1280} height={720} alt="..." />
+```
+
+#### Result
+
+```html
+<!-- Remote images are not optimized or resized. Nonetheless, you still benefit from the enforced attributes and the guarantee of no CLS, albeit manually. -->
+<img
+  src="https://example.com/image.png"
+  decoding="async"
+  loading="lazy"
+  width="1280"
+  height="720"
+  alt="..."
+/>
+```
+
+## Image in Markdown
+
+```markdown
+![My article cover](./cover.png)
+^ Image is 1280x720
+```
+
+### Result
+
+```html
+<!-- Image is optimized, proper attributes are enforced -->
+<img
+  src="/_astro/cover.hash.webp"
+  width="1280"
+  height="720"
+  decoding="async"
+  loading="lazy"
+  alt="My article cover"
+/>
+```
+
+Remote images are handled as in vanilla Markdown and have no special behavior.
+
+## JavaScript API for using optimized images
+
+```astro
+---
+import { getImage } from "astro:assets";
+import image from "../assets/image.png"; // Image is 1920x1080
+
+const myImage = await getImage({ src: image, width: 1280, height: 720 });
+---
+
+<img src={myImage.src} {...myImage.attributes} />
+```
+
+### Result
+
+```html
+<img
+  src="/_astro/image.hash.webp"
+  width="1280"
+  height="720"
+  loading="lazy"
+  decoding="async"
+/>
+```
+
+## Defining a new image service
+
+### Astro Config
+
+```ts
+import { defineConfig } from "astro/config";
+
+// https://astro.build/config
+export default defineConfig({
+  image: {
+    service: "your-entrypoint", // 'astro/image/services/squoosh' | 'astro/image/services/sharp' | string
+  },
+});
+```
+
+### Local services API
+
+```ts
+import type { LocalImageService } from "astro";
```

Comment:
This is a bikeshed comment:

Have you considered using an abstract class for `LocalImageService`?

With an abstract class, we can declare which methods must be implemented and throw errors if these methods are not implemented.

Reply:
`@astrojs/image` actually used classes for this! But it was super confusing for users, since classes are not super common in JavaScript, it just looked like a lot of boilerplate. So, despite my own thoughts on the subject, I went with objects.

I think there might also be some considerations regarding the fact that using classes for this would mean that packages needs a hard dependency on Astro, whereas with the type they can leave it as a dev dep and their package can be leaner if it includes other stuff unrelated to Astro

Reply:
Thank you for providing such background!